src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -5594,3 +5600,33 @@ ExtractParametersFromParamList(ParamListInfo paramListInfo,
 																   parameterData->value);
 	}
 }
+
+
+/*
+ * Delegated functions which are not part of another transaction are naturally safe,
+ * but if it's part of an outer transaction (a 2PC), we should ensure that the current
+ * task will not do remote(connection) activity to ensure the transactional integrity
+ * of the current participating 2PC.
+ */
+static void
+EnsureDelegatedFunctionSafeExecution(List *taskList)
+{
+	if (taskList)
+	{
+		/*
+		 * When the following conditions are met, raise an exception.
+		 * 1) Current process is a remote connection
+		 * 2) In a coordinated distributed transaction
+		 * 3) Distributed transaction (ID) is initiated else where
+		 */
+		DistributedTransactionId *transactionId = GetCurrentDistributedTransactionId();
+
+		if (IsCitusInitiatedRemoteBackend() &&
+			MyBackendIsInDisributedTransaction() &&
+			transactionId->initiatorNodeIdentifier != GetLocalGroupId())
```

Comment:
minor: we typically use early returns to document each individual case, which also forces better explanation / extra thought:
```c
if (taskList == NIL)
{
    /* we allow executions with an empty task list (e.g. WHERE false) */
    returrn;
}

if (!IsCitusInitiatedRemoteBackend())
{
    /* there are no restrictions on application-initiated backends */
    returrn;
}

if (!MyBackendIsInDisributedTransaction())
{
    /* not sure why we need this? */
    returrn;
}

if (transactionId->initiatorNodeIdentifier == GetLocalGroupId())
{
    /* not sure why we need this? */
    returrn;
}

ereport(ERROR, (errmsg("cannot access other workers from operations on shards")));
```

Reply:
Refactored.